### PR TITLE
Add a multiobjective function

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

dagstuhl-request

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/benchmark/test_experiments.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def test_experiments_registry(name: str, maker: Callable[[], Iterator[experiment
     with patch("shutil.which", return_value="here"):  # do not check for missing packages
         with datasets.mocked_data():  # mock mlda data that should be downloaded
             check_maker(maker)  # this is to extract the function for reuse if other external packages need it
-        if "mlda" not in name and "realworld" not in name:
+        if "mlda" not in name and "realworld":
```
Comment (teytaud) [inline nevergrad/benchmark/test_experiments.py]: @jrapin   this change is not clear to me; really ?

File: nevergrad/functions/multiobjective/core.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        # The following is not. It should be call locally.
+        return self.compute_aggregate_loss(losses, *args, **kwargs)
+
+    @property
+    def pareto_front(self) -> List[Tuple[ArgsKwargs, np.ndarray]]:
+        new_points: List[Tuple[ArgsKwargs, np.ndarray]] = []
+        for argskwargs, losses in self._points:
+            for _, other_losses in self._points:
+                if (other_losses <= losses).all() and (other_losses < losses).any():
+                    should_be_added = True
```
Comment (teytaud) [inline nevergrad/functions/multiobjective/core.py]: this line and the line below: something wrong ==> fixed

File: nevergrad/benchmark/test_experiments.py (hunk 1 above)
Comment (jrapin) [reply nevergrad/benchmark/test_experiments.py]: stupid me

File: README.md [hunk 3]
```diff
@@ -94,20 +94,15 @@ Let us minimize x1 and x2 (two objective functions) assuming that values above 2
...
-# The parameter ``instrumentation`` can be replaced by a more sophisticated instrumentation (docs/instrumentation.md).
-# We here just use the dimension as we assume simple continuous rescaled variables (mean 0, std 1).
-optimizer = ng.optimizers.CMA(instrumentation=3, budget=100)  # 2 is the dimension, 100 is the budget.
+optimizer = ng.optimizers.CMA(instrumentation=3, budget=100)  # 3 is the dimension, 100 is the budget.
 recommendation = optimizer.optimize(f)
 
 
 # The function embeds its Pareto-front:
-print("My Pareto front:", f())
+print("My Pareto front:", list(f.pareto_front.values()))
```
Comment (jrapin) [inline README.md]: TODO: pareto_front is a list now

File: nevergrad/benchmark/test_experiments.py [hunk 4]
```diff
@@ -21,7 +21,7 @@ def test_experiments_registry(name: str, maker: Callable[[], Iterator[experiment
     with patch("shutil.which", return_value="here"):  # do not check for missing packages
         with datasets.mocked_data():  # mock mlda data that should be downloaded
             check_maker(maker)  # this is to extract the function for reuse if other external packages need it
-        if "mlda" not in name and "realworld":
+        if "mlda" not in name and "realworld":# not in name:
```
Comment (jrapin) [inline nevergrad/benchmark/test_experiments.py]: TODO: readd not in name

File: nevergrad/functions/multiobjective/core.py [hunk 5]
```diff
@@ -23,13 +23,14 @@ class MultiobjectiveFunction:
...
         self.multiobjective_function = multiobjective_function
+        self.upper_bounds = upper_bounds
         self._hypervolume: Any = _HyperVolume(np.array(upper_bounds))  # type: ignore
         self._points: List[Tuple[ArgsKwargs, np.ndarray]] = []
         self._best_volume = -float("Inf")
 
     def compute_aggregate_loss(self, losses: ArrayLike, *args: Any, **kwargs: Any) -> float:
         # We compute the hypervolume
-        arr_losses = np.array(losses, copy=False)
+        arr_losses = np.minimum(np.array(losses, copy=False), np.array(self.upper_bounds, copy=False))
```
Comment (jrapin) [inline nevergrad/functions/multiobjective/core.py]: TODO: cast upper bounds to array

File: nevergrad/functions/multiobjective/core.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    def __init__(self, multiobjective_function: Callable[..., ArrayLike], upper_bounds: ArrayLike) -> None:
+        self.multiobjective_function = multiobjective_function
+        self._upper_bounds = np.array(upper_bounds, copy=False)
+        self._hypervolume: Any = _HyperVolume(self._upper_bounds)  # type: ignore
+        self._points: List[Tuple[ArgsKwargs, np.ndarray]] = []
+        self._best_volume = -float("Inf")
+
+    def compute_aggregate_loss(self, losses: ArrayLike, *args: Any, **kwargs: Any) -> float:
+        # We compute the hypervolume
+        arr_losses = np.minimum(np.array(losses, copy=False), self._upper_bounds)
```
Comment (jrapin) [inline nevergrad/functions/multiobjective/core.py]: wouldnt it help if we return something like `np.max(np.maximum(np.array(losses, copy=False) - self._upper_bounds, 0))` (i.e. the maximum part above the bounds) when it is strictly positive?

File: nevergrad/functions/multiobjective/test_core.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+
+def test_multiobjective_function() -> None:
+    mfunc = core.MultiobjectiveFunction(lambda x: x, (100, 100))  # type: ignore
+    tuples = [(110, 110), (110, 90), (80, 80), (50, 50), (50, 50), (80, 80), (30, 60), (60, 30)]
+    values = []
+    for tup in tuples:
+        values.append(mfunc(tup))
+    expected = [0, float('inf'), -400, -2500.0, -2500.0, -2500.0, -3300.0, -4100.0]
+    # TODO (80, 80) should yield -2470, and values above the upper bounds could yield positive numbers
```
Comment (jrapin) [inline nevergrad/functions/multiobjective/test_core.py]: Shoulnt the return value for (80,80) be smaller than 2500?

Comment (teytaud) [reply nevergrad/functions/multiobjective/test_core.py]: yes as mentioned in the comment it should be -2470; i've fixed this and others.
an enormous thanks, this test is perfect...

File: nevergrad/functions/multiobjective/core.py (hunk 6 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective/core.py]: yes! done.

File: README.md (hunk 3 above)
Comment (teytaud) [reply README.md]: Done, thx!

**Pull request merged**